Fix animation of stop-color on <stop>

rwlbuis opened the pull request:
#### 3dd3c8cdda3e6451da955890a75c3df9c0a4cb02
<pre>
Fix animation of stop-color on &lt;stop&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=256956">https://bugs.webkit.org/show_bug.cgi?id=256956</a>

Reviewed by Nikolas Zimmermann.

Accept case-insensitive &apos;currentcolor&apos; on color animations.

Tests: imported/w3c/web-platform-tests/svg/animations/animate-stop-currentcolor-ci.html
       imported/w3c/web-platform-tests/svg/animations/animate-stop-currentcolor-escaped.html
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animate-stop-currentcolor-ci-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animate-stop-currentcolor-ci.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animate-stop-currentcolor-escaped-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animate-stop-currentcolor-escaped.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animate-stop-currentcolor-expected.txt:
* Source/WebCore/SaferCPPExpectations/ForwardDeclCheckerExpectations:
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseColorOrCurrentColorWithoutContext):
* Source/WebCore/css/parser/CSSParser.h:
* Source/WebCore/svg/properties/SVGAnimationAdditiveValueFunctionImpl.cpp:
(WebCore::SVGAnimationColorFunction::colorFromString):
* Source/WebCore/svg/properties/SVGPropertyTraits.h:
(WebCore::SVGPropertyTraits&lt;Color&gt;::parse):

Canonical link: <a href="https://commits.webkit.org/302163@main">https://commits.webkit.org/302163@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d8ba3eaf634747f65e923c39b942886a8b71502e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125585 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45247 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35994 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132446 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77478 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127456 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45934 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53807 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95669 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/63559 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128533 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36710 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112303 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76166 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35606 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30485 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75921 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106482 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30703 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135119 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52385 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40136 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104138 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52826 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108519 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103871 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26991 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49222 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27535 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/49588 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52275 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58072 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51629 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54981 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53323 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/svg/properties/SVGAnimationAdditiveValueFunctionImpl.cpp:
```diff
@@ -35,8 +35,9 @@ namespace WebCore {
+    static MainThreadNeverDestroyed<const AtomString> currentcolor("currentcolor"_s);
 
-    if (string != currentColor.get())
+    if (string != currentColor.get() && string != currentcolor.get())
```
r=me.

annevk replied:
Hmm, shouldn't it be case-insensitive then? Add a test for `currentCOLOR` and see what other implementations do?

nt1m replied:
I agree with @annevk

annevk replied:
Note that you can also write `currentc\olor` in CSS. Should escape syntax work here? It being in SVG it might well not be defined I suppose...

rwlbuis replied:
Firefox indeed seems to accept  case-insensitive currentcolor.
Chrome seems to as well, it is hard to say because invalid values also seem to fallback to currentcolor (unlike Firefox).

nt1m replied:
equalLettersIgnoringASCIICase would work here

annevk replied:
We should test the value I mentioned above though. It would be good to confirm there's no need for a full blown CSS parser and then also assert that with a test.

rwlbuis replied:
Indeed the "currentc\olor" case does not work. I wonder if this is the same problem https://bugs.webkit.org/show_bug.cgi?id=42871.

rwlbuis replied:
@annevk how do you think should curre\ntcolor be treated? I do get a feeling the full CSS parser is needed.

annevk replied:
Hey Rob, I think it should be consistent across engines and it should have WPT coverage. I think it's very reasonable for certain SVG attributes to just defer to the CSS parser, similar to what `<canvas>` does. In fact, the more we can share the better. It just needs to be tested in WPT so all implementations end up converging.

rwlbuis replied:
I did another stab at adding tests for this, all three pass.

nikolaszimmermann reviewed (approved):
I thought one AtomicString should be sufficient + using case-insensitive comparison, but I guess you do not want to accept cuRrEnTcolor :-)

annevk reviewed (approved):
Thanks for adding tests!

@weinig might have some thoughts as well, but in principle this looks good to me.

rwlbuis committed:
Fix animation of stop-color on <stop>
https://bugs.webkit.org/show_bug.cgi?id=256956

Reviewed by Nikolas Zimmermann.

Accept case-insensitive 'currentcolor' on color animations.

Tests: imported/w3c/web-platform-tests/svg/animations/animate-stop-currentcolor-ci.html
       imported/w3c/web-platform-tests/svg/animations/animate-stop-currentcolor-escaped.html
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animate-stop-currentcolor-ci-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animate-stop-currentcolor-ci.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animate-stop-currentcolor-escaped-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animate-stop-currentcolor-escaped.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animate-stop-currentcolor-expected.txt:
* Source/WebCore/SaferCPPExpectations/ForwardDeclCheckerExpectations:
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseColorOrCurrentColorWithoutContext):
* Source/WebCore/css/parser/CSSParser.h:
* Source/WebCore/svg/properties/SVGAnimationAdditiveValueFunctionImpl.cpp:
(WebCore::SVGAnimationColorFunction::colorFromString):
* Source/WebCore/svg/properties/SVGPropertyTraits.h:
(WebCore::SVGPropertyTraits<Color>::parse):

Canonical link: https://commits.webkit.org/302163@main
